types/types.go

```diff
@@ -23,6 +24,20 @@ type Repository interface {
 	Image(ref string) Image // ref == image name w/o registry part
 }
 
+// ImageSource is a service, possibly remote (= slow), to download components of a single image.
+type ImageSource interface {
+	GetManifest() (manifest []byte, unverifiedCanonicalDigest string, err error)
+	GetLayer(digest string) (io.ReadCloser, error)
+	GetSignatures() ([][]byte, error)
+}
+
+// ImageDestination is a service, possibly remote (= slow), to store components of a single image.
+type ImageDestination interface {
+	PutManifest([]byte) error
+	PutLayer(digest string, stream io.ReadCloser) error
+	PutSignatures(signatures [][]byte) error
+}
```

Comment:
I think the interface separation is conceptually clearer, and may make a difference if we ever implement direct `docker {save,load}` streaming, where, once the child process is started, it is impossible to reverse the direction of the operation. (`docker {save,load}` might be needed as a component of `atomic {push,pull}`).

Of course a single object can implement both interfaces, so having two different interfaces does not require us to have two different objects for the two directions (though that would require more thought about what exactly the “a single image” is and what the semantics of overwrite are — we can’t overwrite a digest-identified image, so would a single `ImageDestination` refer to a tag? If so, can the contents mutate over time if someone else reassigns the tag?).

At the moment, this is basically a skeleton of an idea: the way the non-`dockerImageSource` code in `dockerImage` is in an awkward limbo clearly indicates the organization of the code needs more thought.

I care little about the naming, what would you prefer?